fix: append `#[::core::prelude::v1::test]` only if it does not exist

kezhuw opened the pull request:
This way if preceding test macros add `#[::core::prelude::v1::test]` by appending, then we can avoid duplicated test runs.

See also frondeus/test-case#101, frondeus/test-case#143

Closes #35.


This pr depends on frondeus/test-case#143 and tokio-rs/tokio#6497.

d-e-s-o commented on macros/src/lib.rs:
```diff
@@ -61,9 +56,24 @@ fn try_test(attr: TokenStream, input: ItemFn) -> syn::Result<Tokens> {
   let tracing_init = expand_tracing_init(&attribute_args);
 
+  let (inner_test, generated_test) = if attr.is_empty() {
+    let test_attr: syn::Attribute = parse_quote! { #[::core::prelude::v1::test] };
```
Type annotation does not seem to be necessary, right?
```suggestion
    let test_attr = parse_quote! { #[::core::prelude::v1::test] };
```

d-e-s-o commented on macros/src/lib.rs:
```diff
@@ -61,9 +56,24 @@ fn try_test(attr: TokenStream, input: ItemFn) -> syn::Result<Tokens> {
+    };
+    (quote! {}, generated_test)
+  } else {
+    let attr: Tokens = attr.into();
```
Personal style, but let's use `From` instead, for the sake of consistency?
```suggestion
    let attr = Tokens::from(attr);
```

d-e-s-o reviewed (approved):
Thanks for the proposal! I am fine with the approach. Left a few nitpicks.

> This pr depends on https://github.com/frondeus/test-case/pull/143 and https://github.com/tokio-rs/tokio/pull/6497.

Can we please wait until these are merged? I'd rather not include any temporary patches.

kezhuw commented on macros/src/lib.rs:
```diff
@@ -61,9 +56,24 @@ fn try_test(attr: TokenStream, input: ItemFn) -> syn::Result<Tokens> {
 
+  let (inner_test, generated_test) = if attr.is_empty() {
+    let test_attr: syn::Attribute = parse_quote! { #[::core::prelude::v1::test] };
+    let has_test = ignored_attrs.iter().any(|attr| *attr == test_attr);
```
It is "tracing-attributes"(through "tracing-subscriber") that enable this for `Display`. But it is gated by feature "trace". I have synced this to what I have done in tokio pr.

kezhuw commented:
@d-e-s-o Thank you for reviewing. I am good with the merge timing.

I have pushed with a fixup commit to solve your concerns. Also, I checked that there is no duplicated test runs for new test cases. `cargo test with_existing` runs `2` tests, `cargo test with_append` runs `7` tests including ones generated by `test_case`.

d-e-s-o commented:
It seems as if the `tokio` change has made it in and `test-case` can't get its act together. @kezhuw I think we use `test-case` only for exercising attribute parsing with special-sauce syntax. If you know of a different crate supporting similar non-standard syntax, then we can just switch existing tests away from `test-case`.

kezhuw commented:
> I think we use test-case only for exercising attribute parsing with special-sauce syntax.

Do you mean tests like this ?

```rust
#[test_log::test(test_case::test_case(-2, -4; "my test name"))]
fn with_inner_test_attribute_and_test_args_and_name(x: i8, y: i8) {
  assert_eq!(x, -2);
  assert_eq!(y, -4);
}
```

This pr does not depend on this. I can drop some tests so that we don't need to patch `test-case` crate.

kezhuw commented:
> I can drop some tests so that we don't need to patch test-case crate.

Those tests should pass once frondeus/test-case#143 published. Or I could `#[ignore]` and doc them.

kezhuw commented:
> we don't want to loose the coverage

For test coverage, the following two methods are identical in `test-log` side except that it will fail without patch to `test-case`. Besides, [`tokio::test`](https://docs.rs/tokio/latest/tokio/attr.test.html) accept

```rust
#[tokio::test]
#[test_log::test]
async fn with_append_test_attribute_and_async() {
  assert_eq!(async { 42 }.await, 42)
}

#[test_case::test_case(-2, -4)]
#[test_case::test_case(-2, -5)]
#[test_log::test]
fn with_append_test_attribute_and_test_args(x: i8, _y: i8) {
  assert_eq!(x, -2);
}
```

> So the easiest way out that I can think of would be to just use some other proc macro that supports more or less arbitrary syntax in test-case's stead.

Let me figure out.

kezhuw commented:
Hi @d-e-s-o, would you mind take another look ? I have replaced `test-case` with `rstest` in commit 628dff9c4e0ea63c7e97d7f622e3ac89129169bf and rebased this pr to main.

d-e-s-o commented on tests/mod.rs:
```diff
@@ -57,6 +57,49 @@ fn with_inner_test_attribute_and_test_args_and_panic(x: i8, _y: i8) {
+#[case(-3, -4)]
+#[test_log::test]
+// Applied to all cases, must not before `rstest`, see https://github.com/la10736/rstest/issues/210
+#[should_panic] // https://docs.rs/rstest/latest/rstest/attr.rstest.html#use-specific-case-attributes
```
Let's use versioned URLs for the sake of persistence? https://docs.rs/rstest/0.25.0/rstest/attr.rstest.html#use-specific-case-attributes

d-e-s-o commented on tests/mod.rs:
```diff
@@ -57,6 +57,49 @@ fn with_inner_test_attribute_and_test_args_and_panic(x: i8, _y: i8) {
+#[case(-2, -4)]
+#[case(-3, -4)]
+#[test_log::test]
+// Applied to all cases, must not before `rstest`, see https://github.com/la10736/rstest/issues/210
```
```suggestion
// Applied to all cases, must not come before `rstest`, see https://github.com/la10736/rstest/issues/210
```

d-e-s-o replied:
Same below.

d-e-s-o commented on tests/mod.rs:
```diff
@@ -5,12 +5,12 @@
 use tracing::info;
 use tracing::instrument;
 
-
```
Nit: Let's keep spurious changes to a minimum?

d-e-s-o reviewed (commented):
I think this looks great, thanks! Left a few more comments. Your initial PR description mentioned that we *depend* on https://github.com/frondeus/test-case/pull/143. Why is that no longer the case? Unfortunately somehow GitHub seems to have managed to loose the original set of changes... :-(

kezhuw commented:
Here is the no rebased branch: https://github.com/d-e-s-o/test-log/commits/628dff9c4e0ea63c7e97d7f622e3ac89129169bf/

I think we are here because of https://github.com/d-e-s-o/test-log/pull/46#issuecomment-2587168118.

d-e-s-o reviewed (approved):
Looks good to me. Feel free to squash the fixup commit and then we can get it in. Thanks again for your work.

kezhuw commented:
Squashed. @d-e-s-o Thank you for reviewing!
